pyvisfile/xdmf/__init__.py

```diff
@@ -0,0 +1,1119 @@
+
+    def as_data_item(self, *,
+            parent: Optional[Element] = None) -> Tuple[DataItem, ...]:
+        items = super().as_data_item(parent=parent)
+
+        if self.ary.dtype.char == "O":
+            ary = tuple(self.ary)
+        else:
+            ary = (self.ary,)
+
+        for item, iary in zip(items, ary):
+            item.text = _ndarray_to_string(iary)
+
+        return items
+
+# }}}
+
+
+# {{{ grids
+
+class XdmfGrid:
+    def __init__(self, root):
+        self.root = root
+
+    def getroot(self):
+        return self.root
+
+    def add_attribute(self, ary: DataArray, *, join=True) -> Attribute:
+        acenter = ary.acenter
+        if acenter is None:
+            acenter = AttributeCenter.Node
+
+        atype = ary.atype
+        if atype is None:
+            atype = _attribute_type_from_shape(ary.shape[::-1])
+
+        attr = Attribute(
+                name=ary.name,
+                atype=atype,
+                acenter=acenter,
+                parent=self.getroot(),
+                )
+
+        if join:
+            items = ary.as_data_item()
+            join_data_items(items, parent=attr)
+        else:
+            ary.as_data_item(parent=attr)
+
+        return attr
+
+
+class XdmfUnstructuredGrid(XdmfGrid):
+    def __init__(self,
+            points: DataArray,
+            connectivity: DataArray, *,
+            topology_type: Union[Topology, TopologyType],
+            name: Optional[str] = None,
+            geometry_type: Optional[GeometryType] = None):
+        if geometry_type is None:
+            geometry_type = _geometry_type_from_points(points)
+
+        if geometry_type not in (GeometryType.XY, GeometryType.XYZ):
+            # NOTE: Paraview 5.8 seems confused when using VXVY geometry types
+            raise ValueError(f"unsupported geometry type: '{geometry_type}'")
+
+        grid = Grid(parent=None, name=name)
+
+        nelements = np.prod(connectivity.shape[:-1])
+        if isinstance(topology_type, TopologyType):
+            if topology_type == TopologyType.Polyline:
+                nodes_per_element = 2
+            else:
+                nodes_per_element = None
+
+            topology = Topology(
+                    parent=grid,
+                    ttype=topology_type,
+                    nodes_per_element=nodes_per_element,
+                    number_of_elements=nelements)
+        elif isinstance(topology, Topology):
+            topology = topology_type.replace({
+                "parent": grid,
+                "number_of_elements": nelements
+                })
+        else:
+            raise TypeError(f"unsupported type: {type(topology_type).__name__}")
+
+        connectivity.as_data_item(parent=topology)
+
+        geometry = Geometry(parent=grid, gtype=geometry_type)
+        points.as_data_item(parent=geometry)
+
+        super().__init__(grid)
+
+
+class XdmfStructuredGrid(XdmfGrid):
+    def __init__(self):
+        super().__init__()
```

Comment:
Is this unsupported as yet? (If so, maybe raise `NotImplementedError`?) Or is this already the whole implementation?

Reply:
It's not implemented, no. Just removed it in d85ad44.